File: server_elixir/test/afterlight/activities/foosball_test.exs
defmodule Afterlight.Activities.FoosballTest do
  use ExUnit.Case, async: true
  alias Afterlight.Activities.Foosball

  describe "initialization and dimensions" do
    test "table dimensions and goal specifications" do
      assert Foosball.length() == 120.0
      assert Foosball.width() == 70.0
      assert Foosball.goal_width() == 20.0
      {g_top, g_bottom} = Foosball.goal_bounds()
      assert g_top == 25.0
      assert g_bottom == 45.0
    end

    test "default simulation state initializes correctly" do
      state = Foosball.init_sim_state()
      assert state["width"] == 70
      assert state["length"] == 120
      assert state["state"] == "serving"
      assert state["score"] == %{"0" => 0, "1" => 0}
      assert state["seriesScore"] == %{"0" => 0, "1" => 0}
      assert state["targetScore"] == 5
      assert state["seriesLength"] == 1
      assert state["winsNeeded"] == 1
      assert is_map(state["rods"])
      assert length(state["rods"]["0"]) == 4
      assert length(state["rods"]["1"]) == 4
    end

    test "series length selection respects permitted lengths" do
      assert Foosball.wins_needed(1) == 1
      assert Foosball.wins_needed(3) == 2
      assert Foosball.wins_needed(5) == 3
      assert Foosball.wins_needed(7) == 1 # clamped to 1 if not in [1, 3, 5]

      s3 = Foosball.init_sim_state(series_length: 3)
      assert s3["seriesLength"] == 3
      assert s3["winsNeeded"] == 2
    end
  end

  describe "rod constraints and angular velocity limits" do
    test "rod lateral translation is clamped within bounds" do
      rod = %{"y" => 35.0, "angle" => 0.0, "vy" => 0.0, "omega" => 0.0}
      # Goalie bounds for slot 0: min 25.0, max 45.0
      clamped_low = Foosball.clamp_rod_state(0, 0, rod, %{"targetY" => 10.0})
      assert clamped_low["y"] >= 25.0

      clamped_high = Foosball.clamp_rod_state(0, 0, rod, %{"targetY" => 60.0})
      assert clamped_high["y"] <= 45.0
    end

    test "rod angular velocity is strictly bounded to <= 15 rad/s" do
      rod = %{"y" => 35.0, "angle" => 0.0, "vy" => 0.0, "omega" => 0.0}
      # Attempt massive instant angular jump
      clamped = Foosball.clamp_rod_state(0, 1, rod, %{"targetAngle" => 3.14159})
      max_omega = Foosball.max_angular_speed()
      assert abs(clamped["omega"]) <= max_omega + 1.0e-5
      assert abs(clamped["angle"]) <= max_omega + 1.0e-5
    end

    test "unrestricted 360-degree spinning is prohibited" do
      rod = %{"y" => 35.0, "angle" => 1.5, "vy" => 0.0, "omega" => 0.0}
      clamped = Foosball.clamp_rod_state(0, 2, rod, %{"targetAngle" => 20.0})
      max_angle = Foosball.max_rod_angle()
      assert clamped["angle"] <= max_angle
      assert clamped["angle"] >= -max_angle
    end

    test "casual mode recommends the active rod by ball position" do
      assert Foosball.get_recommended_rod(0, 15.0) == 0 # Goalie
      assert Foosball.get_recommended_rod(0, 35.0) == 1 # Defense
      assert Foosball.get_recommended_rod(0, 60.0) == 2 # Midfield
      assert Foosball.get_recommended_rod(0, 85.0) == 3 # Attack

      assert Foosball.get_recommended_rod(1, 110.0) == 0 # Goalie
      assert Foosball.get_recommended_rod(1, 85.0) == 1  # Defense
      assert Foosball.get_recommended_rod(1, 60.0) == 2  # Midfield
      assert Foosball.get_recommended_rod(1, 35.0) == 3  # Attack
    end
  end

  describe "ball physics, rails, and corner ramps" do
    test "ball bounces off side rails" do
      state =
        Foosball.init_sim_state()
        |> Map.put("state", "rally")
        |> put_in(["ball", "x"], 60.0)
        |> put_in(["ball", "y"], 2.1)
        |> put_in(["ball", "vx"], 0.0)
        |> put_in(["ball", "vy"], -5.0)

      {stepped, _} = Foosball.step(state, %{}, 1)
      assert stepped["ball"]["vy"] > 0.0
      assert stepped["ball"]["y"] >= 2.0
    end

    test "ball deflects off corner ramps preventing dead corners" do
      state =
        Foosball.init_sim_state()
        |> Map.put("state", "rally")
        |> put_in(["ball", "x"], 3.0)
        |> put_in(["ball", "y"], 3.0)
        |> put_in(["ball", "vx"], -4.0)
        |> put_in(["ball", "vy"], -4.0)

      {stepped, _} = Foosball.step(state, %{}, 1)
      # Velocity should bounce back towards the positive quadrant
      assert stepped["ball"]["vx"] > 0.0 or stepped["ball"]["vy"] > 0.0
    end

    test "ball kicks upon player figure contact when kicking" do
      # Place ball just in front of P0 Midfield rod (x = 66.0, centered figure at y = 35.0)
      state =
        Foosball.init_sim_state()
        |> Map.put("state", "rally")
        |> put_in(["ball", "x"], 68.0)
        |> put_in(["ball", "y"], 35.0)
        |> put_in(["ball", "vx"], 0.0)
        |> put_in(["ball", "vy"], 0.0)

      players = %{
        0 => %{input_state: %{"kick" => true, "selectRod" => 2, "controlMode" => "advanced"}}
      }

      {stepped, _} = Foosball.step(state, players, 1)
      # Kicking rod should drive ball forward (+X)
      assert stepped["ball"]["vx"] > 0.0
    end
  end

  describe "goals, double-goal lock, and match conclusion" do
    test "goal scored for slot 1 when ball crosses left goal mouth" do
      state =
        Foosball.init_sim_state()
        |> Map.put("state", "rally")
        |> put_in(["ball", "x"], 2.0)
        |> put_in(["ball", "y"], 35.0)
        |> put_in(["ball", "vx"], -10.0)
        |> put_in(["ball", "vy"], 0.0)

      {stepped, outcome} = Foosball.step(state, %{}, 1)
      assert stepped["state"] == "goal"
      assert stepped["score"]["1"] == 1
      assert stepped["score"]["0"] == 0
      assert stepped["goalDelay"] == 60
      assert outcome == nil
    end

    test "goal scored for slot 0 when ball crosses right goal mouth" do
      state =
        Foosball.init_sim_state()
        |> Map.put("state", "rally")
        |> put_in(["ball", "x"], 118.0)
        |> put_in(["ball", "y"], 35.0)
        |> put_in(["ball", "vx"], 10.0)
        |> put_in(["ball", "vy"], 0.0)

      {stepped, outcome} = Foosball.step(state, %{}, 1)
      assert stepped["state"] == "goal"
      assert stepped["score"]["0"] == 1
      assert stepped["score"]["1"] == 0
      assert stepped["goalDelay"] == 60
      assert outcome == nil
    end

    test "ball crossing end line outside goal mouth bounces off end rail" do
      # End rail above goal mouth: y = 10.0 (goal is 25.0..45.0)
      state =
        Foosball.init_sim_state()
        |> Map.put("state", "rally")
        |> put_in(["ball", "x"], 2.0)
        |> put_in(["ball", "y"], 10.0)
        |> put_in(["ball", "vx"], -8.0)
        |> put_in(["ball", "vy"], 0.0)

      {stepped, _} = Foosball.step(state, %{}, 1)
      assert stepped["state"] == "rally"
      assert stepped["score"]["1"] == 0
      assert stepped["ball"]["vx"] > 0.0
    end

    test "double-goal rejection: goals are ignored during goal delay" do
      state =
        Foosball.init_sim_state()
        |> Map.put("state", "goal")
        |> Map.put("goalDelay", 30)
        |> Map.put("score", %{"0" => 1, "1" => 0})
        |> put_in(["ball", "x"], 1.0)
        |> put_in(["ball", "y"], 35.0)
        |> put_in(["ball", "vx"], -10.0)

      {stepped, _} = Foosball.step(state, %{}, 1)
      assert stepped["score"]["1"] == 0
      assert stepped["score"]["0"] == 1
      assert stepped["goalDelay"] == 29
    end

    test "first-to-five wins the game and concludes single-game match" do
      state =
        Foosball.init_sim_state(series_length: 1)
        |> Map.put("state", "goal")
        |> Map.put("goalDelay", 1)
        |> Map.put("score", %{"0" => 5, "1" => 3})

      {stepped, outcome} = Foosball.step(state, %{}, 1)
      assert stepped["state"] == "ended"
      assert stepped["winner"] == 0
      assert stepped["seriesScore"]["0"] == 1
      assert match?({:match_ended, 0, _}, outcome)
    end

    test "best-of-three series transitions to game_break when game won but series ongoing" do
      state =
        Foosball.init_sim_state(series_length: 3)
        |> Map.put("state", "goal")
        |> Map.put("goalDelay", 1)
        |> Map.put("score", %{"0" => 5, "1" => 2})
        |> Map.put("seriesScore", %{"0" => 0, "1" => 0})

      {stepped, outcome} = Foosball.step(state, %{}, 1)
      assert stepped["state"] == "game_break"
      assert stepped["currentGame"] == 2
      assert stepped["seriesScore"]["0"] == 1
      assert outcome == nil
    end
  end
end

File: tests/foosball.test.js
import test from 'node:test';
import assert from 'node:assert/strict';
import {
  TABLE_LENGTH,
  TABLE_WIDTH,
  CENTER_X,
  CENTER_Y,
  BALL_RADIUS,
  GOAL_WIDTH,
  GOAL_TOP,
  GOAL_BOTTOM,
  ROD_CONFIGS,
  initFoosballState,
  winsNeededForSeries,
  getRecommendedRod,
  clampRodState,
  stepFoosball,
} from '../shared/foosballModel.js';

test('1. Foosball table dimensions, constants, and rod configurations', () => {
  assert.equal(TABLE_LENGTH, 120.0);
  assert.equal(TABLE_WIDTH, 70.0);
  assert.equal(CENTER_X, 60.0);
  assert.equal(CENTER_Y, 35.0);
  assert.equal(BALL_RADIUS, 2.0);
  assert.equal(GOAL_WIDTH, 20.0);
  assert.equal(GOAL_TOP, 25.0);
  assert.equal(GOAL_BOTTOM, 45.0);

  // Both slots have 4 rods
  assert.equal(ROD_CONFIGS['0'].length, 4);
  assert.equal(ROD_CONFIGS['1'].length, 4);

  // Total 8 rods across the table
  const allRods = [...ROD_CONFIGS['0'], ...ROD_CONFIGS['1']];
  assert.equal(allRods.length, 8);

  // Check rod figures count: 1 + 2 + 5 + 3 = 11 figures per team
  const p0TotalFigs = ROD_CONFIGS['0'].reduce((sum, r) => sum + r.playerCount, 0);
  const p1TotalFigs = ROD_CONFIGS['1'].reduce((sum, r) => sum + r.playerCount, 0);
  assert.equal(p0TotalFigs, 11);
  assert.equal(p1TotalFigs, 11);
});

test('2. Foosball state initialization and series selection', () => {
  assert.equal(winsNeededForSeries(1), 1);
  assert.equal(winsNeededForSeries(3), 2);
  assert.equal(winsNeededForSeries(5), 3);
  assert.equal(winsNeededForSeries(7), 1); // Not in [1, 3, 5], defaults to 1

  const sDefault = initFoosballState();
  assert.equal(sDefault.length, 120);
  assert.equal(sDefault.width, 70);
  assert.equal(sDefault.state, 'serving');
  assert.equal(sDefault.score['0'], 0);
  assert.equal(sDefault.score['1'], 0);
  assert.equal(sDefault.targetScore, 5);
  assert.equal(sDefault.seriesLength, 1);
  assert.equal(sDefault.winsNeeded, 1);

  const sBo3 = initFoosballState({ seriesLength: 3 });
  assert.equal(sBo3.seriesLength, 3);
  assert.equal(sBo3.winsNeeded, 2);
});

test('3. Rod translation clamping and angular velocity bounds (no unrestricted spinning)', () => {
  const currentRod = { y: 35.0, angle: 0.0, vy: 0.0, omega: 0.0 };

  // 1. Translation clamping: Goalie minY=25, maxY=45
  const clampedLow = clampRodState(0, 0, currentRod, { targetY: 5.0 });
  assert.ok(clampedLow.y >= 25.0);

  const clampedHigh = clampRodState(0, 0, currentRod, { targetY: 65.0 });
  assert.ok(clampedHigh.y <= 45.0);

  // 2. Angular speed limit (<= 15 rad/s -> ~0.25 rad/tick)
  const clampedAngle = clampRodState(0, 1, currentRod, { targetAngle: 3.0 });
  assert.ok(Math.abs(clampedAngle.omega) <= (15.0 / 60.0) + 1e-5);
  assert.ok(Math.abs(clampedAngle.angle) <= (15.0 / 60.0) + 1e-5);

  // 3. Prohibit 360-degree continuous spinning: angle clamped to [-0.48*PI, 0.48*PI]
  const extremeSpin = clampRodState(0, 2, { y: 35, angle: 1.5, vy: 0, omega: 0 }, { targetAngle: 100.0 });
  assert.ok(extremeSpin.angle <= Math.PI * 0.48);
  assert.ok(extremeSpin.angle >= -Math.PI * 0.48);
});

test('4. Casual active rod recommendation and advanced manual selection', () => {
  // Slot 0 (attacks right)
  assert.equal(getRecommendedRod(0, 15.0), 0); // Goalie
  assert.equal(getRecommendedRod(0, 35.0), 1); // Defense
  assert.equal(getRecommendedRod(0, 60.0), 2); // Midfield
  assert.equal(getRecommendedRod(0, 85.0), 3); // Attack

  // Slot 1 (attacks left)
  assert.equal(getRecommendedRod(1, 110.0), 0); // Goalie
  assert.equal(getRecommendedRod(1, 85.0), 1);  // Defense
  assert.equal(getRecommendedRod(1, 60.0), 2);  // Midfield
  assert.equal(getRecommendedRod(1, 35.0), 3);  // Attack

  // In stepFoosball with casual mode, active rod tracks ball automatically
  let state = initFoosballState();
  state.state = 'rally';
  state.ball.x = 85.0; // In slot 0's attack zone
  const res = stepFoosball(state, {});
  assert.equal(res.state.activeRod['0'], 3); // Automatically chose attack rod!

  // In advanced mode, explicit selection overrides
  const advRes = stepFoosball(state, {
    '0': { controlMode: 'advanced', selectRod: 1 },
  });
  assert.equal(advRes.state.controlMode['0'], 'advanced');
  assert.equal(advRes.state.activeRod['0'], 1);
});

test('5. Corner ramps, rails, and figure strikes', () => {
  let state = initFoosballState();
  state.state = 'rally';
  state.ball.x = 4.0;
  state.ball.y = 4.0;
  state.ball.vx = -4.0;
  state.ball.vy = -4.0;

  // Corner ramp deflection
  const { state: cornerBounced } = stepFoosball(state, {});
  assert.ok(cornerBounced.ball.vx > 0 || cornerBounced.ball.vy > 0, 'corner ramp deflected ball away from corner');

  // Side rail bounce
  let railState = initFoosballState();
  railState.state = 'rally';
  railState.ball.x = 60.0;
  railState.ball.y = 2.0;
  railState.ball.vx = 0.0;
  railState.ball.vy = -5.0;
  const { state: railBounced, events: railEvents } = stepFoosball(railState, {});
  assert.ok(railBounced.ball.vy > 0, 'bounced off top rail');
  assert.ok(railEvents.some(e => e.type === 'bounce_rail'));

  // Figure strike upon kicking
  let kickState = initFoosballState();
  kickState.state = 'rally';
  kickState.ball.x = 68.0;
  kickState.ball.y = 35.0;
  kickState.ball.vx = 0.0;
  kickState.ball.vy = 0.0;

  const { state: kickedState, events: kickEvents } = stepFoosball(kickState, {
    '0': { controlMode: 'advanced', selectRod: 2, kick: true },
  });
  assert.ok(kickedState.ball.vx > 0, 'figure strike propelled ball forward');
  assert.ok(kickEvents.some(e => e.type === 'strike'));
});

test('6. Goals, double-goal lock, and first-to-five match conclusion', () => {
  // 1. Goal scored for slot 1 (left goal)
  let sGoal1 = initFoosballState();
  sGoal1.state = 'rally';
  sGoal1.ball.x = 2.0;
  sGoal1.ball.y = 35.0;
  sGoal1.ball.vx = -10.0;
  sGoal1.ball.vy = 0.0;

  const { state: postGoal1, events: goal1Events } = stepFoosball(sGoal1, {});
  assert.equal(postGoal1.state, 'goal');
  assert.equal(postGoal1.score['1'], 1);
  assert.equal(postGoal1.score['0'], 0);
  assert.equal(postGoal1.goalDelay, 60);
  assert.ok(goal1Events.some(e => e.type === 'goal' && e.scorer === 1));

  // 2. Double-goal rejection during goal delay
  const { state: postLock } = stepFoosball(postGoal1, {});
  assert.equal(postLock.score['1'], 1, 'score does not increment again during lock');
  assert.equal(postLock.goalDelay, 59);

  // 3. Goal scored for slot 0 (right goal)
  let sGoal0 = initFoosballState();
  sGoal0.state = 'rally';
  sGoal0.ball.x = 118.0;
  sGoal0.ball.y = 35.0;
  sGoal0.ball.vx = 10.0;
  sGoal0.ball.vy = 0.0;

  const { state: postGoal0 } = stepFoosball(sGoal0, {});
  assert.equal(postGoal0.score['0'], 1);
  assert.equal(postGoal0.state, 'goal');

  // 4. First-to-five ends single game match
  let sEnd = initFoosballState({ seriesLength: 1 });
  sEnd.state = 'goal';
  sEnd.goalDelay = 1;
  sEnd.score = { '0': 5, '1': 3 };

  const { state: matchEnded, events: endEvents } = stepFoosball(sEnd, {});
  assert.equal(matchEnded.state, 'ended');
  assert.equal(matchEnded.winner, 0);
  assert.equal(matchEnded.seriesScore['0'], 1);
  assert.ok(endEvents.some(e => e.type === 'match_won' && e.winner === 0));
});
